doc/developer/design/20230306_persist_mfp_pushdown.md

```diff
@@ -0,0 +1,272 @@
+- Feature name: Persist Filter Pushdown
+- Associated: https://github.com/MaterializeInc/materialize/issues/12684
+
+# Summary
+[summary]: #summary
+
+Queries against large collections in Materialize can be very expensive:
+readers generally need to fetch all the data in the collection as of the query time,
+even if they're only interested in a small subset.
+By pushing down filters into the Persist source,
+we can dramatically improve performance for certain subsets of queries,
+especially those that filter down to a narrow window of time.
+
+# Motivation
+[motivation]: #motivation
+
+Unlock append-only sources for customers via temporal filters. Right now, using an append-only `SOURCE` in Materialize is painful because `MATERIALIZED VIEW` restart times increase without bound as data accumulates. However, if there is a [temporal filter](https://materialize.com/docs/sql/patterns/temporal-filters/) to bound the size of the dataflow's input data (e.g. the last 30 days of data), we may be able to use the described optimization to similarly bound the restart times.
+
+This technique is not limited to temporal filters or sources. In fact, the optimization will be picked up anywhere we can apply it. However, we're hoping to focus on polishing the subset relevant to append-only sources in the first version of this shipped to customers.
+
+# Explanation
+[explanation]: #explanation
+
+## Schemas
+
+A `Schema` trait will be introduced to map data back and forth between Persist columns. Any Rust type
+that is to roundtrip through Persist must have a corresponding `Schema`.
+
+Each column in a `RelationDesc` is mapped to a column that persist understands: a name, a type, and override-able logic for computing aggregate statistics of columns of that data. The set of supported types is a subset of Apache [Arrow's data types], which is then internally mapped to [Parquet].
+
+[Arrow's data types]: https://docs.rs/arrow2/0.17.0/arrow2/datatypes/enum.DataType.html
+[parquet]: https://parquet.apache.org/docs/
+## Part Statistics
+
+When Persist writes a batch to Blob storage, it will compute basic statistics over its columns, when possible,
+and write this data alongside existing metadata in Consensus. To start, we will collect a column's
+non-null min and max values and null count. We may also track the min/max Row within the part.
+
+Statistics are optional: Persist may not be able to compute min/max values for all types, and we may need to trim statistics to keep metadata from getting too big.
```

Comment:
> we may need to trim statistics to keep metadata from getting too big

Any ideas which policies are going to govern how those statistics will be trimmed beyond the initial "restrict our attention to temporal filters" approach? Other databases allow users to specify the set of columns for which statistics should be available.

Reply:
this is covered in a bit more depth in the Reference section, so I added a pointer to it from here. i agree that we'll probably want to let users control this at some point, but i think that's probably out of the initial scope of the work, so i intentionally left the specifics vague ("we almost certainly want some way for the customer to be able to hint that a certain column is critical and should never have trimmed stats")